datalad/distribution/subdatasets.py

```diff
@@ -268,7 +268,7 @@ def __call__(
                         k)
         try:
             if not (bottomup or contains or set_property or delete_property or \
-                    (recursive and recursion_limit is not None)):
+                    recursive):
                 # FAST IMPLEMENTATION FOR THE STRAIGHTFORWARD CASE
```

Comment:
This will have undesirable sideeffects. With this change every subdataset will get a separate call to `git diff`, instead of a single call to `git diff` for the entire tree. This is a substantial slowdown, as `git diff` will now evaluate the state of each submodule twice. Seperate calls are only necessary to support behavior that `git diff` cannot handle directly, such as recursion limits, reverse order, and path matches.

Reply:
this logic is far from my comprehension so unlikely I would be able to deliver the "correct" fix here.